nmnb-application/src/main/kotlin/nmnb/application/global/auth/service/RefreshTokenService.kt

```diff
@@ -1,38 +1,74 @@
 package nmnb.application.global.auth.service
 
-import nmnb.application.global.auth.exception.AuthException
 import nmnb.application.global.infrastructure.security.JwtTokenProvider
+import nmnb.common.response.exception.AuthException
 import nmnb.common.response.status.ErrorStatus
 import nmnb.domain.auth.RefreshToken
 import nmnb.domain.auth.repository.RefreshTokenRepository
 import org.springframework.data.repository.findByIdOrNull
 import org.springframework.stereotype.Component
+import java.time.LocalDateTime
 
 @Component
 class RefreshTokenService(
     private val refreshTokenRepository: RefreshTokenRepository,
     private val tokenProvider: JwtTokenProvider,
 ) {
-    fun validateRefreshToken(refreshToken: String): String {
+    fun validateRefreshToken(refreshToken: String, deviceId: String): String {
         val email = tokenProvider.getEmailWithValidation(refreshToken)
-        verifyStoredTokenMatch(email, refreshToken)
+        verifyStoredTokenMatch("$email:$deviceId", refreshToken)
         return email
     }
 
-    fun verifyStoredTokenMatch(email: String, token: String) {
-        val storedToken = refreshTokenRepository.findByIdOrNull(email)?.refreshToken
+    fun verifyStoredTokenMatch(id: String, token: String) {
+        val storedToken = refreshTokenRepository.findByIdOrNull(id)?.refreshToken
 
         if (storedToken != token) {
             throw AuthException(ErrorStatus.AUTH_INVALID_TOKEN)
         }
     }
 
-    fun saveOrUpdateToken(email: String, refreshToken: String) {
-        val existingToken = refreshTokenRepository.findByIdOrNull(email)
+    fun upsertRefreshToken(email: String, deviceId: String, refreshToken: String) {
+        val tokenId = "$email:$deviceId"
+        val existingToken = refreshTokenRepository.findByIdOrNull(tokenId)
+
+        val now = LocalDateTime.now()
+
         if (existingToken != null) {
-            existingToken.update(refreshToken)
+            val updated = existingToken.copy(
+                refreshToken = refreshToken,
+                timeStamp = now,
+            )
+            refreshTokenRepository.save(updated)
         } else {
-            refreshTokenRepository.save(RefreshToken(email, refreshToken))
+            val newToken = RefreshToken(
+                id = tokenId,
+                email = email,
+                deviceId = deviceId,
+                refreshToken = refreshToken,
+                timeStamp = now,
+            )
+            refreshTokenRepository.save(newToken)
         }
     }
+
+    fun removeOldestTokenIfLimitExceeded(email: String) {
+        val allTokens = getUserRefreshTokensSortedByTime(email)
+
+        if (allTokens.size <= MAX_REFRESH_TOKENS) return
+
+        val tokensToRemove = allTokens
+            .take(allTokens.size - MAX_REFRESH_TOKENS)
+
+        tokensToRemove.forEach { refreshTokenRepository.deleteById(it.id) }
```

Comment:
이 코드 지금도 좋지만 조금 더 간결하게 쓸 수 있을 것 같아요!!
`tokens.size <= MAX_REFRESH_TOKENS`인 경우 return으로 빠져나오는 것보다 return문을 빼고 조건문을 반대로 구성하면 흐름이 더 깔끔해질 것 같다는 생각이 들었습니다!
```suggestion
val allTokens = getUserRefreshTokensSortedByTime(email)

if (allTokens.size > MAX_REFRESH_TOKENS) {
    val tokensToRemove = allTokens.take(allTokens.size - MAX_REFRESH_TOKENS)
    tokensToRemove.forEach { refreshTokenRepository.deleteById(it.id) }
}
```
그리고` allTokens.take(allTokens.size - MAX_REFRESH_TOKENS)` 부분을 조금 더 이해하기 쉽게 바꿀 수 있지 않을까 싶어서 알아보니까 **뒤에서 N개를 남기고 앞부분을 제거**하는 `dropLast()`라는 함수가 있더라구요. 최신 N개를 남기고 오래된 토큰 제거하는 의도를 잘 전달하고 있는 함수인 것 같아서 다음과 같이 바꾸는 것은 어떨지도 제안해봅니다!

```kotlin
val tokensToRemove = allTokens.dropLast(MAX_REFRESH_TOKENS)
```

Reply:
1) guard clause 패턴을 사용해서 조건을 만족하지 않을 때 조기 반환하고, 나머지 코드를 평평하게 유지하는 방식으로 코드 작성했습니다! 이 방식이 Kotlin 스타일에서도 권장되는 패턴이라 가독성 측면에서 이점이 있다고 판단했습니다.

=> if 조건을 반대로 작성해서 감싸는 것도 가능하지만, 그럴 경우 로직이 깊어질 수 있어 현재 구조가 더 간결하다고 생각합니다! 이 부분은 그냥 그대로 두는거 어떨지 제안합니다~!

Reply:
2) dropLast로 변경하는거 좋은거 같아요 ! 좀 더 직관적이고 의도가 명확한거 같네요 수정하겠습니다!

Reply:
오오 권장되는 패턴이었군요!! 저도 이후 로직 작성할 때 참고하겠습니다! 
그대로 두는 것 좋습니다!